File: go/g2d/actor.go
package g2d

type Actor interface {
    Move()
    Collide(other Actor)
    Position() Point
    Size() Point
    Symbol() Point
}

type Arena struct {
    w, h   int
    actors []Actor
}

func NewArena(size Point) *Arena {
    a := &Arena{size.X, size.Y, []Actor{}}
    return a
}

func (a *Arena) Add(actor Actor) {
    if !a.Contains(actor) {
        a.actors = append(a.actors, actor)
    }
}

func (a *Arena) Remove(actor Actor) {
    for i, v := range a.actors {
        if v == actor {
            a.actors = append(a.actors[:i], a.actors[i+1:]...)
            return
        }
    }
}

func (a *Arena) Contains(actor Actor) bool {
    for _, v := range a.actors {
        if v == actor {
            return true
        }
    }
    return false
}

func (a *Arena) MoveAll() {
    actors := a.ReversedActors()
    for _, actor := range actors {
        actor.Move()
        for _, other := range actors {
            if actor != other && a.CheckCollision(actor, other) {
                actor.Collide(other)
                other.Collide(actor)
            }
        }
    }
}

func (a *Arena) CheckCollision(a1, a2 Actor) bool {
    p1 := a1.Position()
    s1 := a1.Size()
    p2 := a2.Position()
    s2 := a2.Size()
    return (p2.X < p1.X+s1.X && p1.X < p2.X+s2.X &&
        p2.Y < p1.Y+s1.Y && p1.Y < p2.Y+s2.Y &&
        a.Contains(a1) && a.Contains(a2))
}

func (a *Arena) Actors() []Actor {
    actors := make([]Actor, len(a.actors))
    for i, v := range a.actors {
        actors[i] = v
    }
    return actors
}

func (a *Arena) ReversedActors() []Actor {
    actors := make([]Actor, len(a.actors))
    for i, v := range a.actors {
        actors[len(a.actors)-i-1] = v
    }
    return actors
}

func (a *Arena) Size() Point {
    return Point{a.w, a.h}
}

File: cpp/bounce/actor.cpp
#include "actor.h"

Arena::Arena(int width, int height) {
    w_ = width;
    h_ = height;
}

void Arena::add(Actor* a) {
    auto pos = find(begin(actors_), end(actors_), a);
    if (pos == actors_.end()) {
        actors_.push_back(a);
    }
}

void Arena::remove(Actor* a) {
    auto pos = find(begin(actors_), end(actors_), a);
    if (pos != actors_.end()) {
        actors_.erase(pos);
    }
}

void Arena::move_all() {
    auto acts = actors();
    reverse(begin(acts), end(acts));
    for (auto a : acts) {
        auto prev = a->rect();
        a->move();
        auto curr = a->rect();
        if (curr != prev) {
            for (auto other : acts) {
                if (other != a && check_collision(a, other)) {
                    a->collide(other);
                    other->collide(a);
                }
            }
        }
    }
}

bool Arena::check_collision(Actor* a1, Actor* a2) {
    auto r1 = a1->rect();
    auto r2 = a2->rect();
    return (r2[1] < r1[1] + r1[3] && r1[1] < r2[1] + r2[3]
        && r2[0] < r1[0] + r1[2] and r1[0] < r2[0] + r2[2]);
}

vector<Actor*> Arena::actors() { return actors_; }

vector<int> Arena::size() { return {w_, h_}; }

Arena::~Arena() {
    while (!actors_.empty()) {
        delete actors_.back();
        actors_.pop_back();
    }
}
